dbms/src/Storages/KVStore/Utils/AsyncTasks.h

```diff
@@ -14,110 +14,358 @@
 
 #pragma once
 
+#include <Common/FmtUtils.h>
+#include <Common/Logger.h>
+#include <Common/SyncPoint/SyncPoint.h>
 #include <Common/UniThreadPool.h>
 
 #include <future>
+#include <magic_enum.hpp>
+#include <unordered_map>
 
 namespace DB
 {
+
+namespace AsyncTaskHelper
+{
+// It's guaranteed a task is no longer accessible once canceled or has its result fetched.
+// NotScheduled -> InQueue, Running, NotScheduled
+// InQueue -> Running
+// Running -> Finished, NotScheduled(canceled)
+// Finished -> NotScheduled(fetched)
+// NOTE: magic_enum can't work properly with enums in template classes due to compiler opts.
+enum class TaskState
+{
+    NotScheduled,
+    InQueue,
+    Running,
+    Finished,
+};
+} // namespace AsyncTaskHelper
+
 template <typename Key, typename Func, typename R>
 struct AsyncTasks
 {
     // We use a big queue to cache, to reduce add task failures.
     explicit AsyncTasks(uint64_t pool_size, uint64_t free_pool_size, uint64_t queue_size)
         : thread_pool(std::make_unique<ThreadPool>(pool_size, free_pool_size, queue_size))
+        , log(DB::Logger::get("AsyncTasks"))
```

Comment:
```suggestion
        , log(DB::Logger::get())
```
Adding "AsyncTasks" for it is useless. The logging already contain the file and lineno. Usually we use the identifier to track a "task". For example, to filter the logging for a given table_id, mpp_id, or apply snapshot for a given region_id